File: Trees/Diameter of a BT.py
  #TC:O(n)
  #SC:O(height)
  #Function to return the diameter of a Binary Tree.
    def diameter(self,root):
        maxLength = float('-inf') 
        
        #diameterHelper returns the maxlength from root node to any of the leaf nodes reachable from that root node.
        
        def diameterHelper(root):
            nonlocal maxLength
            if root == None:return 0
            
            #Max length from its left and right childs to their leaf nodes is calculated
            leftMaxLength = diameterHelper(root.left)
            rightMaxLength = diameterHelper(root.right)
            
            #Comparing Maxlength with the maximum length path passing through this particular root node
            maxLength = max(maxLength,1+leftMaxLength+rightMaxLength)
            
            
            return 1+max(leftMaxLength,rightMaxLength)
        
        diameterHelper(root)
        
        return maxLength

File: Trees/Diameter of a BT.java
  #TC:O(n)
  #SC:O(height)
    
  #Function to return the diameter of a Binary Tree.
  class Solution {
    // Function to return the diameter of a Binary Tree.
    static int maxLength = Integer.MIN_VALUE;
    int diameter(Node root) {
        diameterHelper(root);
        //System.out.println(maxLength);
        return maxLength;
    }
    int diameterHelper(Node root){
        if(root==null)return 0;
        int lh = diameterHelper(root.left);
        int rh = diameterHelper(root.right);
        maxLength = Math.max(maxLength,1+lh+rh);
        //System.out.println(maxLength);
        return 1 + Math.max(lh,rh);
    }
}
